web: Add StatusMessageItem to NavBar

mrnugget opened the pull request:
This is the second part of the implementation of #4120, the UI, as described in https://github.com/sourcegraph/sourcegraph/issues/4120#issuecomment-499911568. (It's a stacked PR that depends on #4440 being merged first)

This adds the `StatusMessagesNavItem` to the global `NavBar`. Read through #4120 to get a general idea of the feature and what this MVP entails.

Here is what it looks like when nothing is currently being cloned/updated:

<img width="540" alt="Screen Shot 2019-06-11 at 15 26 36" src="https://user-images.githubusercontent.com/1185253/59276038-d383fb00-8c5d-11e9-9636-94a7f0f6bc38.png">

Here is what it currently looks like when repositories are being cloned (or updated/fetched!) in the background:

<img width="541" alt="Screen Shot 2019-06-11 at 15 27 00" src="https://user-images.githubusercontent.com/1185253/59276073-e3034400-8c5d-11e9-9e1e-47dba4035668.png">

What I'm still unsure about:
1. Maybe it makes sense to add a header to the `DropdownMenu`? The goal for this item is to be expanded in the future and a header is a chance to give more context, like this maybe?
<img width="383" alt="Screen Shot 2019-06-11 at 15 02 30" src="https://user-images.githubusercontent.com/1185253/59276160-0e862e80-8c5e-11e9-83af-c3bfb6fe0a61.png">
2. The wording, here and in #4440. Since the spinner activates whenever we do the initial clone of **OR** just update a repository, maybe "cloning" is wrong and we should change it to "updating"? Technically it is cloning, but it happens far more often than the initial cloning

Reviewers: I'm totally new to the `web` codebase and its conventions, so, fire away :)

Test plan: jest tests

mrnugget commented:
FYI: I want to land this feature in 3.5 on Friday, so the sooner you can give me your _actionable_ feedback, the better :)

lguychard commented on web/src/nav/StatusMessageNavItem.test.tsx:
```diff
@@ -0,0 +1,13 @@
+    setLinkComponent((props: any) => <a {...props} />)
+    afterAll(() => setLinkComponent(null as any)) // reset global env for other tests
+
+    test('nothing cloning', () => {
```
You should add snapshot tests for the "cloning" state as well

mrnugget replied:
Done!

mrnugget commented:
Question about the Percy tests: will this result in ultra-flaky builds if I approve the difference due to the spinner? As I understand it, if the screenshot is taken at a slightly different time, the spinner looks different and the diff will be different, which leads the build to fail.

lguychard commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,97 @@
+    isOpen: boolean
+}
+
+const refreshInternalMs = 3000
```
```suggestion
const REFRESH_INTERVAL_MS = 3 * 1000
```

lguychard commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,97 @@
+    public state: State = { isOpen: false, messages: [] }
+
+    public componentDidMount(): void {
+        this.subscriptions.add(
```
You could get rid of the `setTimeout()` and the `notificationUpdates` Subject with the following:

```typescript
timer(0, REFRESH_INTERVAL_MS).pipe(
    concatMap(() => fetchAllStatusMessages())
).subscribe(messages => this.setState({ messages }))
```

mrnugget replied:
This is the _good stuff_ I've been looking for, yiss

lguychard commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,97 @@
+    public render(): JSX.Element | null {
+        const types = this.state.messages.map(n => n.type)
+        const hasMessages = this.state.messages.length > 0
+        const currentlyCloning = types.some(t => t === GQL.StatusMessageType.CURRENTLYCLONING)
```
```suggestion
        const currentlyCloning = this.state.messages.some(({ type }) => type === GQL.StatusMessageType.CURRENTLYCLONING)
```

lguychard commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,97 @@
+
+const refreshInternalMs = 3000
+
+export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
```
Please add a docstring for this component.

lguychard commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,97 @@
+}
+
+interface Props {
+    messages?: GQL.IStatusMessage[]
```
This doesn't seem to be used?

mrnugget replied:
Yeah, I just used it to add another test that snapshots the component _with_ messages being rendered

felixfbecker replied:
It's odd to have this both from Props and fetched - instead you would take the `fetchAllStatusMessages` function as a prop, and then pass different ones in tests that just return constant data. That also allows you to test the error case.

mrnugget replied:
If I do this, how do I get the `timer` thing to kick off in the tests? It seems like it doesn't run in the snapshot tests

lguychard replied:
Your `Props` would look like this:

```typescript
interface Props {
    fetchAllStatusMessages: () => Observable<GQL.IStatusMessage[]>
}
```

Which you could inject as follows in tests (obviously the data here is bogus/wrong):

```typescript
// No messages
<StatusMessagesNavItem fetchAllStatusMessages={() => of([])}/>

// One message
<StatusMessagesNavItem fetchAllStatusMessages={() => of(['cloning'])}/>

// Error case
<StatusMessagesNavItem fetchAllStatusMessages={() => throwError(new Error('no'))}/>
```

mrnugget replied:
As discussed on Slack, that's what I did in https://github.com/sourcegraph/sourcegraph/pull/4441/commits/a43eea5a2a12b118e70f5fcd481aa7b49a67fdce and https://github.com/sourcegraph/sourcegraph/pull/4441/commits/74ddbbe52732d2a2991709ed3e5119362adc1ec5 but this doesn't work for me, even when passing `asapScheduler` to the `timer`.

I'm quite out of my depth here :)

felixfbecker replied:
Pushed a commit to pass in a scheduler (`queueScheduler`, which is synchronous). `asapScheduler` is actually scheduling on the next possible tick

lguychard commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,97 @@
+        this.subscriptions.unsubscribe()
+    }
+
+    private renderMessage(message: GQL.IStatusMessage): JSX.Element | null {
```
This doesn't need to be extracted into its own method.

mrnugget replied:
That's right and it does look a little sparse, but I think that it makes `render` itself more readable. Is there any argument against extracting things solely for readability purposes?

Ideally, once we have more than one type of message, I'd turn this `renderMessage` into a `StatusMessage` component

lguychard replied:
I'm OK with leaving this as-is until then.

felixfbecker replied:
When adding other `render*` methods to a component it is usually a red flag that this should be put in another small function component. The fact that all CSS class names in here are prefixed with `status-messages-nav-item__entry-` is also a red flag that `StatusMessagesNavItemEntry` makes sense as its own component.

lguychard commented on web/src/nav/StatusMessagesNavItem.scss:
```diff
@@ -0,0 +1,5 @@
+.status-messages-nav-item__dropdown-menu {
```
```suggestion
.status-messages-nav-item {
    &__dropdown-menu {
```

lguychard commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,97 @@
+    isOpen: boolean
+}
+
+const refreshIntervalMs = 3000
```
```suggestion
const REFRESH_INTERVAL_MS = 3000
```

lguychard commented:
> Question about the Percy tests: will this result in ultra-flaky builds if I approve the difference due to the spinner?

Since we control when the Percy snapshots are taken (with `percySnapshot()`), I would ensure that the repos are cloned before taking a snapshot.

mrnugget commented:
> Since we control when the Percy snapshots are taken (with `percySnapshot()`), I would ensure that the repos are cloned before taking a snapshot.

Hmmm, I see. The problem is that we cannot really control this, since cloning/updating/git-fetching happens periodically in the background.

mrnugget commented:
Another idea for the "no messages" dropdown:

<img width="533" alt="Screen Shot 2019-06-11 at 17 05 00" src="https://user-images.githubusercontent.com/1185253/59283577-13051400-8c6b-11e9-973d-cd49895ec900.png">

tsenart commented:
> Another idea for the "no messages" dropdown:

Needs a green check mark :)

tsenart commented:
I'm very impressed with your frontend skillzzzz. :muscle:

mrnugget commented:
> I'm very impressed with your frontend skillzzzz. 💪

I'm printing this out and put it in a frame. This goes up on my wall. You're the first person to have ever said this 🙇‍♂️

felixfbecker commented on web/src/nav/StatusMessageNavItem.test.tsx:
```diff
@@ -0,0 +1,22 @@
+
+describe('UserNavItem', () => {
+    setLinkComponent((props: any) => <a {...props} />)
+    afterAll(() => setLinkComponent(null as any)) // reset global env for other tests
```
Why is `as any` needed?

mrnugget replied:
Straight up copied from the existing `UserNavItem` test (from which I also copied the `describe`). Both fixed.

mrnugget replied:
<img width="1001" alt="Screen Shot 2019-06-11 at 19 11 29" src="https://user-images.githubusercontent.com/1185253/59292154-bf032b00-8c7c-11e9-9b11-a78fcf358718.png">

That's why it's needed I guess

felixfbecker commented:
I find the checkmark icon and loader don't communicate well that this is about repos, the button has no label and there is a somewhat similar status icon for the setup project. Can we maybe find a different icon that creates an association with downloads or repos?

mrnugget commented:
> Can we maybe find a different icon that creates an association with downloads or repos?

Agree! It does look generic. What we need is something that tells us "OK" plus "external" plus "connection" or something.

felixfbecker commented:
Check https://materialdesignicons.com/
Also consider you could stack icons :)

mrnugget commented:
> Also consider you could stack icons :)

Duuude! TIL about [`<Stack>`](https://www.npmjs.com/package/@mdi/react#stack-usage). Nice.

christinaforney commented:
## Headers/Empty state
I like the more detailed empty state screenshot you added - it feels like a right first step in teaching users that they can come to here to understand the state of connection to their services. 

**Empty state wording:**
For both empty and active dropdowns, the "Configure external services" should only be shown to users that can actually configure an external service (I know this iteration we are only showing this icon to those people, but want to make sure we don't forget to do this when we represent state to all users).

> Repositories up to date
>
> All repositories hosted on the configured external services are up to date.
>
> Configure external services

**Active state wording:**
Can we differentiate between the initial cloning state and an updating state? If so, I would like to reflect that in the wording, because it has a different expectation if we are doing a `git pull` vs. `git clone`. Additionally, I'm worried showing an active state while updating could become too much - how often do we update? In reality, are we always updating? If so, updating should not be an active state as indicated by the top level icon, and instead could be reflected by the content of the dropdown.

> Repositories (cloning|updating)
>
> Currently (cloning|updating) N repositories...
>
> Configure external services

## Icons

### Consideration
Related to @tsenart 's comment about a green checkmark in the success/empty state -if we have multiple services connected, perhaps we want to show state for each, as represented by different icons. In this case, we need to be able to capture global state in a toolbar icon (but this can lead to icon overload). 

Instead of these icons we could simply have the text 'Status' as the menu prompt. Thinking about making sure this is accessible, icons alone are often ambiguous and need to be learned.

Related reading on icon accessibility:
[Stack exchange suggestion regarding AA standards](https://ux.stackexchange.com/questions/96163/icon-only-vs-text-and-icon)
[Accessible Web Design: The visual-only icon problem (a short article)](https://www.joedolson.com/2014/06/visual-icon-problem/)
[Longer detailed discussion summarized by previous link](https://www.filamentgroup.com/lab/bulletproof_icon_fonts.html)

### Icon options

Not sure if "cloud" is the right symbol, but a few icon options (this would remove the animation of in progress, which removes the similarity to the Setup progress):
https://materialdesignicons.com/icon/cloud-check 
https://materialdesignicons.com/icon/cloud-sync
https://materialdesignicons.com/icon/cloud-alert
![image](https://user-images.githubusercontent.com/1559622/59304381-feb21e80-8c4c-11e9-82c3-708a9ccb1528.png) ![image](https://user-images.githubusercontent.com/1559622/59304398-0bcf0d80-8c4d-11e9-9241-96ebb8b22470.png) ![image](https://user-images.githubusercontent.com/1559622/59304425-19849300-8c4d-11e9-8e43-3ff89180ef5a.png)

Possible options that are "code" related (possibly wrap them in a circle? we'd have to come up with an in-progress state)
https://materialdesignicons.com/icon/code-tags
https://materialdesignicons.com/icon/code-tags-check
![image](https://user-images.githubusercontent.com/1559622/59304475-30c38080-8c4d-11e9-8757-15eb70748df7.png) ![image](https://user-images.githubusercontent.com/1559622/59304497-3d47d900-8c4d-11e9-9ad6-9f115abff7c9.png)

Another source code related that we would have to stack/color for identifying state:
https://materialdesignicons.com/icon/git
![image](https://user-images.githubusercontent.com/1559622/59304341-e9d58b00-8c4c-11e9-9f75-da18226dc74d.png)

sqs commented:
Can we consolidate this with the `Setup` item? Currently that is used for the initial onboarding flow. But in a sense it communicates the same thing as StatusMessageItem: "You are in a global state that is maybe not ideal, and here's what that means or how to improve it".

This UI element essentially becomes our status bar. In the future :thinking:...

- It could start to hold things like the Sentry/LightStep/Datadog/Codecov enablement status:, which would declutter the file header.
- It could hold notifications for checks and codemods.

Here's an idea. It is underspecified; let me know if it would help for me to think through this in more detail.

- Consolidate both `Setup` and StatusMessageItem into a single global nav item whose icon is a circle.
- We define the possible states of the circle, like:
  - a loading/not-loading state (BTW spinning things on screen make me dizzy, as a user I would not want this to be spinning for a long time and globally visible)
  - a numeric indicator
  - a look-at-me/severity state
- We define the possible entries of the menu, using a very simple schema (text + link + tooltip + faded state + loading state) and jam both the current `Setup` items and this new cloning item into the menu
- If there is anything in the menu that we want to make sure the user sees (eg cloning), we can programmatically open the menu (similar to how Google Cloud opens the operations menu the first time you perform a long-running operation).

The hope is that this would be the 1 place users are trained to look if "something seems wrong" and there is no contextual error. This is similar to how VS Code extensions will show an `X` for broken code intel, linting, Live Share connection, Git syncing, etc., in the status bar, which is where users expect to look. This works acceptably despite it not being focus-grabbing because it's a single place that users need to look.

![image](https://user-images.githubusercontent.com/1976/59333600-8891d400-8cad-11e9-806e-9f782b5b2cae.png)

felixfbecker commented:
I like the cloud option @christinaforney proposed because the cloud icon is also used on the repository page to represent cloning status. I think not having it be animated can actually be a good thing to not make it too distracting (couldn't it otherwise be spinning _a lot_ with a lot of repos & users?).
I think the icon should not need to represent individual states but only the aggregate status (and don't users usually not have tons of external services, but more commonly just one?).

keegancsmith commented:
Just touching web, so will leave code review to others. As for the above discussion, it's important to note we want this to extend to conveying all sorts of background task information (not just related to repos). So @sqs idea of consolidating is a good one.

However, I also worry about the amount of debate we are having. It would be great to agree on something we can land now so we can start providing value to users in the next release. We can tweak / overhaul in follow-up PRs if needs be. So if this debate is still continuing on Thursday, I suggest we ensure whatever the current state of the PR is is merged in (so please also review the current technical implementation, even if it will change).

mrnugget commented:
> Can we consolidate this with the Setup item? 
> [...]

@sqs I really like this idea! But the short answer to your question is: not in this release, no. Two reasons for that:
1. with my UI/frontend skills as they are I personally don't feel I can pull of a merge of both UI elements without high-fidelity mockups
2. we have 3 days left until the release branch for `3.5` is cut and afterwards we're quite busy with (load-)testing the release. I don't want to cut into this time by spending a lot of time _this week_ on testing/QA of a merged setup-progress/clone-status-indicator element

I totally agree with @keegancsmith in that I think the best way forward would be to try to see how we can make the best version of what we agreed upon in scope (see original ticket) so that we can land it in this release and already provide insight. In the next release we can extend this, merge it, add new info to it, etc.

sqs commented:
Sounds good. :smile:

mrnugget commented:
### Update after incorporating feedback
#### Screenshots

Inactive:
<img width="535" alt="second_version_inactive" src="https://user-images.githubusercontent.com/1185253/59352672-85d0c680-8d21-11e9-81c6-14376ff4ac55.png">

Active:
<img width="457" alt="Screen Shot 2019-06-12 at 14 51 25" src="https://user-images.githubusercontent.com/1185253/59352707-93864c00-8d21-11e9-9d05-3ed8c90c8759.png">

GIF:

![second_iteration_animated](https://user-images.githubusercontent.com/1185253/59352714-97b26980-8d21-11e9-84b1-ffd8014fe274.gif)

#### Changes

Okay, here's what I did:
* Incorporated technical feedback
* Changed to static cloud icons — I really like these and now that I've used it a bunch of times, I have to say that non-spinner icon is probably the way to go. If it has to be animated, it needs to be really, really subtle
* Changed the active/inactive look of the dropdown to both have a header/body/link
* Give both icon versions a tooltip on hover that hides on click (especially proud of that detail 😎)
* Changed the number in the message (see https://github.com/sourcegraph/sourcegraph/pull/4440#issuecomment-501236329 for details) — @christinaforney I think for now (this release) we can't easily determine between "cloning" and "updating". The icon does switch to "active" on a regular basis (whenever we poll the external services in the background, every 10min, IIRC, and when we run a scheduled update of a repository) but IMHO without a spinner it doesn't feel like too much or distracting, because it's a pretty subtle change in the icon 

#### Next steps

After now changing to the number to "enqueued" (which gives a great sense of feedback) and switching to the non-animated cloud icons (which give context and are rather subtle) I now feel confident in shipping _this_.

What do you think @christinaforney?

Before I hit merge, I still want to hide this icon in Percy though, since the automatic updates in the background could run at any time and cause Percy flakyness...

And, please, @felixfbecker and @lguychard, if you have the time, take another look at the changes I made

felixfbecker commented:
Please wait for my review

felixfbecker commented on web/src/nav/StatusMessagesNavItem.scss:
```diff
@@ -0,0 +1,19 @@
+.status-messages-nav-item {
+    &__nav-link {
+        @media only percy {
+            visibility: hidden;
```
I am worried that this circumvents the point of Percy - this means if this status indicator regresses, we won't catch it. Why do @lguychard's suggestion of ensuring repos are cloned before running tests / taking snapshots? I would think we need to do that anyway?

mrnugget replied:
> Why do @lguychard's suggestion of ensuring repos are cloned before running tests / taking snapshots? I would think we need to do that anyway?

Yeah, we already do that:

https://github.com/sourcegraph/sourcegraph/blame/master/web/src/e2e/index.e2e.test.tsx#L58-L61

Should we bump it to 3 minutes to be sure? Or 4? 😄

There's two reasons why I want to take this icon out of Percy:
1. At the moment I don't think we have a programmatic way to determine when all repositories are cloned. We could increase the wait time again to make sure _all_ repositories are cloned, but that would... well, make the e2e tests even slower for an icon that
2. might change at any moment anyway, because we're polling/updating the repositories in the background a) on a regular basis and b) depending on how the frequency of updates to the repositories

I don't think the benefit of detecting regressions in the icon outweigh the potential flakiness of the tests.

felixfbecker commented on web/src/nav/StatusMessagesNavItem.test.tsx:
```diff
@@ -0,0 +1,32 @@
+import { StatusMessagesNavItem } from './StatusMessagesNavItem'
+
+describe('StatusMessagesNavItem', () => {
+    setLinkComponent((props: any) => <a {...props} />)
```
I don't get a compile error when removing the `any`, so it seems unnecessary 
```suggestion
    setLinkComponent(props => <a {...props} />)
```

felixfbecker commented on web/src/nav/StatusMessagesNavItem.test.tsx:
```diff
@@ -0,0 +1,32 @@
+
+describe('StatusMessagesNavItem', () => {
+    setLinkComponent((props: any) => <a {...props} />)
+    afterAll(() => setLinkComponent(null as any)) // reset global env for other tests
```
```suggestion
    afterAll(() => setLinkComponent(() => null)) // reset global env for other tests
```

This behaves the same in my testing and is type safe

felixfbecker commented on web/src/nav/StatusMessagesNavItem.test.tsx:
```diff
@@ -0,0 +1,32 @@
+    test('no messages', () => {
+        expect(renderer.create(<StatusMessagesNavItem messages={[]} />).toJSON()).toMatchSnapshot()
+    })
+    test('one CLONING message - not site admin', () => {
```
Instead of duplicating "one CLONING message", nest these two tests in another `describe('one CLONING message')`.
That also gives you a scope to share the duplicated `message` variable.

felixfbecker commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,129 @@
+function fetchAllStatusMessages(): Observable<GQL.IStatusMessage[]> {
+    return queryGraphQL(
+        gql`
+            query {
```
Add a query name so it shows up in the network panel

felixfbecker commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,129 @@
+
+        this.subscriptions.add(
+            timer(0, REFRESH_INTERVAL_MS)
+                .pipe(concatMap(fetchAllStatusMessages))
```
Careful, `fetchAllStatusMessages` does a network request so prone to error. That would terminate the stream here and break the UI. Instead, the error should be caught and saved in state, then displayed in the UI with an alert (while still retrying). There are many examples of this in other components (see state named `fooOrError` and usage of `catchError()` operator)

mrnugget replied:
I tried this:

```typescript
        this.subscriptions.add(
            timer(0, REFRESH_INTERVAL_MS)
                .pipe(
                    catchError(err => {
                        console.error(err)
                        return []
                    }),
                    concatMap(() => this.props.fetchMessages())
                )
                .subscribe(messages => this.setState({ messages }))
        )
```

But that only catches the error and logs it. It seems to stop the timer though?

felixfbecker replied:
Yeah the `catchError()` here is applied before the `fetchMessages()` (i.e. it would catch errors from `timer()`, which never errors). You need to catch errors from `fetchMessages()`:

```ts
 this.subscriptions.add(
            timer(0, REFRESH_INTERVAL_MS)
                .pipe(
                    concatMap(() => this.props.fetchMessages().pipe(
                        catchError(err => [asError(err)])
                    ))
                )
                .subscribe(messagesOrError => this.setState({ messagesOrError }))
        )
```

Then in `render()` you use `isErrorLike()` in a ternary to display a different icon and the error message in an `alert alert-danger` in the popover.

felixfbecker commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,129 @@
+                        )}
+                    </div>
+                )
+            default:
```
Should this have a default branch? Without a default branch, TypeScript should give you a compile error if you don't handle all enum values, which can make extending this in the future & ensuring all are handled easier

felixfbecker commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,129 @@
+                    {cloning ? (
+                        <CloudSyncIcon
+                            className="icon-inline"
+                            {...(!this.state.isOpen && { 'data-tooltip': 'Updating repositories...' })}
```
This looks odd, why the spread syntax of an object literal and not a conditional value?

```suggestion
                            data-tooltip={this.state.isOpen ? undefined: 'Updating repositories...'}
```

felixfbecker commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,129 @@
+
+                <DropdownMenu right={true} className="status-messages-nav-item__dropdown-menu">
+                    {hasMessages ? (
+                        this.state.messages.map(this.renderMessage.bind(this))
```
Prefer arrow functions over `bind()`
```suggestion
                        this.state.messages.map(m => this.renderMessage(m))
```

mrnugget replied:
Out of curiosity, since I grew up on `.bind()` and always saw proper usage of `bind` combined with passing-functions-as-values as a sign that someone understood `this` in JavaScript — why?

felixfbecker replied:
`bind` is notoriously slow. I think there was some effort at some point to optimize it in Chrome but it shouldn't be generally assumed to be fast in all browsers. It is also not type safe in previous TypeScript versions and the type checking is pretty complex in newer versions (error messages won't be very clear). There is also a pitfall when the function takes multiple parameters (or may in the future) since `map` passes the index as the second argument and array as third.

An arrow function on the other hand is fast (has no prototype, and usually gets inlined), safe this, and is always type safe.

The only argument I have _for_ `bind` is code coverage, if the arrow function is impossible to test.

mrnugget replied:
Ah, interesting, TIL :) Thanks!

felixfbecker commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,129 @@
+                            </p>
+                            {this.props.isSiteAdmin && (
+                                <p className="status-messages-nav-item__entry-link">
+                                    <Link to={'/site-admin/external-services'}>Configure external services</Link>
```
```suggestion
                                    <Link to="/site-admin/external-services">Configure external services</Link>
```

felixfbecker commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -0,0 +1,129 @@
+                    ) : (
+                        <div className="status-messages-nav-item__entry">
+                            <h4 className="status-messages-nav-item__entry-title">Repositories up to date</h4>
+                            <p className="status-messages-nav-item__entry-copy">
```
I don't see these CSS classes being referenced

felixfbecker commented:
Visual feedback:
- The left and right padding looks to big
- The outline of the icon doesn't use our outline style - which indicates it's not using `<button class="btn btn-icon">`

christinaforney commented:
### Responses for this release
> * Changed to static cloud icons — I really like these and now that I've used it a bunch of times, I have to say that non-spinner icon is probably the way to go. If it has to be animated, it needs to be really, really subtle
> * Changed the active/inactive look of the dropdown to both have a header/body/link

These look great! 

> * Give both icon versions a tooltip on hover that hides on click (especially proud of that detail 😎)

VERY nice touch 🥇 

> * Changed the number in the message (see [#4440 (comment)](https://github.com/sourcegraph/sourcegraph/pull/4440#issuecomment-501236329) for details) — @christinaforney I think for now (this release) we can't easily determine between "cloning" and "updating". The icon does switch to "active" on a regular basis (whenever we poll the external services in the background, every 10min, IIRC, and when we run a scheduled update of a repository) but IMHO without a spinner it doesn't feel like too much or distracting, because it's a pretty subtle change in the icon

Agree! I like this a LOT better. Totally fine if we don't have cloning vs. updating different this time around, especially with the changes you introduced.

> #### Next steps
> After now changing to the number to "enqueued" (which gives a great sense of feedback) and switching to the non-animated cloud icons (which give context and are rather subtle) I now feel confident in shipping _this_.
> 
> What do you think @christinaforney?

Yes! This is a great first version and I am comfortable with this going into the release.

Response to @sqs design feedback coming soon!

mrnugget commented:
> Visual feedback:
> 
> * The left and right padding looks to big
> * The outline of the icon doesn't use our outline style - which indicates it's not using `<button class="btn btn-icon">`

Cleaned up the classes I used, removed horizontal padding, looks much better:

<img width="358" alt="Screen Shot 2019-06-12 at 16 32 14" src="https://user-images.githubusercontent.com/1185253/59359984-a7d14580-8d2f-11e9-91ea-ac7c13cc5bdf.png">

Problem is that using `nav={true}` for the `DropdownToggle` gives it a `btn-group` which gives it a `display: inline-flex` which makes the vertical alignment slightly off. But I also don't want to introduce a special `margin-top: -4px` or something like that. Maybe someone who actually knows what they're doing can go through the navbar and fix the vertical alignment of all nav items?

tsenart commented:
@mrnugget: Others are more qualified to review this code, so removed myself as a reviewer.

mrnugget commented on web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -125,8 +128,10 @@ export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
+                    {isErrorLike(this.state.messagesOrError) ? (
+                        <div className="alert alert-danger">{startCase(this.state.messagesOrError.message)}</div>
+                    ) : this.state.messagesOrError.length > 0 ? (
+                        this.state.messagesOrError.map(m => this.renderMessage(m))
```
Holy ternary operator 😮

christinaforney commented:
Continuing discussion about future of the system status nav item here: #4519
